open-apparel-registry/open-apparel-registry
# Enable Users to create API Authorization Tokens

## Overview

This PR enables users to create API Authorization Tokens through the UI. In prep for that I restored a not-currently-editable version of the User Profile page (shown only to the logged in user), added GET, POST, and DELETE endpoints for the DRF Auth Token, and added an "API Tokens" section to the profile page with the ability to get, copy, delete, and generate a new API token.

Connects #75

## Demo

![auth-token](https://user-images.githubusercontent.com/4165523/51561698-ebe1ad80-1e55-11e9-90ac-6afe3146937b.gif)

![screen shot 2019-01-22 at 2 57 10 pm](https://user-images.githubusercontent.com/4165523/51561746-04ea5e80-1e56-11e9-917d-ccf5c43cd97b.png)

## Notes

As described in #119 multiple-tokens-per-user isn't directly supported by DRF, although there are several viable workarounds. The client was set up in such a way as to afford the possibility of having multiple tokens in case we decide to implement #119.

## Testing

- get this branch, then `./scripts/update` and `./scripts/server`
- sign up with a new user through the UI, then log in with that user
- on your user profile page, verify that you see the "API Tokens" section beneath the user data form
- use the API tokens section
    - generate a new token
    - copy a token to the clipboard
    - refresh the page and verify that the token is fetched
    - delete a token
- generate a token, then in the VM run the following command with your token:

```sh
http http://localhost:8081/token-auth-example/ 'Authorization: Token <YOUR-TOKEN>'
```

...and verify that that returns 200 with a JSON response including your user name.

- try the following command and verify that it returns 401/unauthorized:

```sh
http http://localhost:8081/token-auth-example/ 'Authorization: Token hello-world'
```
